h5py/tests/test_group.py

```diff
@@ -697,6 +697,48 @@ def test_bailout(self):
         x = self.f.visititems(lambda x, y: (x,y))
         self.assertEqual(x, (self.groups[0], self.f[self.groups[0]]))
 
+class TestVisitLinks(TestCase):
+    """
+        Feature: The .visit_links and .visititems_links methods allow iterative access to
+        links contained in the group and its subgroups.
+    """
+
+    def setUp(self):
+        self.f = File(self.mktemp(), 'w')
+        self.groups = [
+            'grp1', 'grp1/grp11', 'grp1/grp12', 'grp2', 'grp2/grp21', 'grp2/grp21/grp211'
+            ]
+        self.links = [
+            'linkto_grp1', 'grp1/linkto_grp11', 'grp1/linkto_grp12', 'linkto_grp2', 'grp2/linkto_grp21', 'grp2/grp21/linkto_grp211'
+        ]
+        for g, l in zip(self.groups, self.links):
+            self.f.create_group(g)
+            self.f[l] = SoftLink(f'/{g}')
+
+    def tearDown(self):
+        self.f.close()
+
+    def test_visit_links(self):
+        """ All subgroups and links are visited """
+        l = []
+        self.f.visit_links(l.append)
+        self.assertSameElements(l, self.groups + self.links)
+
+    def test_visititems(self):
+        """ All links are visited """
+        l = []
+        comp = [(x, type(self.f.get(x, getlink=True))) for x in self.groups + self.links]
+        self.f.visititems_links(lambda x, y: l.append((x, type(y))))
```

Comment:
I resorted to test for the type of links because apparently it is not possible to test for link equality??

```python
>>> l = f.get('grp1', getlink=True)
>>> l2 = f.get('grp1', getlink=True)
>>> l == l2
False
```

:thinking:

Reply:
That might be easy to fix, but it raises the question: if two different files contain links with the same path, which would point to different objects because of starting at different places, do we expect them to be equal?

At present, the Python `SoftLink` & `ExternalLink` classes just store the path & (for external links) the filename. They don't keep any reference to the file they came from.

My leaning is that that's OK, two links with the same path can be equal, even if within their files they would resolve differently. :shrug:

Reply:
Well, I guess it depends on your usecase.

Personally, I can't think of a usecase where you would want to check for link equality, except when testing. :shrug: